libraries/chain/include/eosio/chain/backing_store/kv_context_rocksdb.hpp

```diff
@@ -0,0 +1,451 @@
+            return kv_it_stat::iterator_ok;
+         } else {
+            actual_size = 0;
+            return kv_it_stat::iterator_end;
+         }
+      }
+
+    private:
+      void get_current_key_value_sizes(uint32_t* found_key_size, uint32_t* found_value_size) {
+         if (kv_it_status() == kv_it_stat::iterator_ok) {
+            // Not end or at an erased kv pair
+            auto kv = *kv_current;
+            *found_value_size = backing_store::actual_value_size(
+                  kv.second->size()); // This must be before *found_key_size in case actual_value_size throws
+            *found_key_size = actual_key_size(kv.first.size());
+         } else {
+            *found_key_size   = 0;
+            *found_value_size = 0;
+         }
+      }
+   }; // kv_iterator_rocksdb
+
+   template <typename Session, typename Resource_manager>
+   struct kv_context_rocksdb : kv_context {
+      using session_type = std::remove_pointer_t<Session>;
+
+      session_type*                               session;
+      name                                        receiver;
+      Resource_manager                            resource_manager;
+      const kv_database_config&                   limits;
+      uint32_t                                    num_iterators = 0;
+      eosio::session::shared_bytes                current_key;
+      std::optional<eosio::session::shared_bytes> current_value;
+
+      kv_context_rocksdb(session_type& the_session, name receiver, Resource_manager /*kv_resource_manager*/ resource_manager,
+                         const kv_database_config& limits)
+          : session{ &the_session }, receiver{ receiver },
+            resource_manager{ resource_manager }, limits{ limits } {}
+
+      int64_t kv_erase(uint64_t contract, const char* key, uint32_t key_size) override {
+         EOS_ASSERT(name{ contract } == receiver, table_operation_not_permitted, "Can not write to this key");
+         
+         current_value.reset();
+         current_key                = eosio::session::shared_bytes{};
+         auto old_value             = std::optional<eosio::session::shared_bytes>{};
+         auto actual_old_value_size = uint32_t{ 0 };
+         try {
+            try {
+               auto composite_key = make_composite_key(contract, nullptr, 0, key, key_size);
+               old_value          = session->read(composite_key);
+               if (!old_value)
+                  return 0;
+               actual_old_value_size = backing_store::actual_value_size(old_value->size());
+               session->erase(composite_key);
+            }
+            FC_LOG_AND_RETHROW()
+         }
+         CATCH_AND_EXIT_DB_FAILURE()
+
+         account_name payer = backing_store::get_payer(old_value->data());
+         return erase_table_usage(resource_manager, payer, key, key_size, actual_old_value_size);
+      }
+
+      int64_t kv_set(uint64_t contract, const char* key, uint32_t key_size, const char* value, uint32_t value_size,
```

Comment:
The kv_context_chainbase logs to deep mind.

Reply:
Added